Source/MQTTnet/Client/MqttPacketDispatcher.cs

```diff
@@ -8,18 +9,29 @@ namespace MQTTnet.Client
     public class MqttPacketDispatcher
     {
         private readonly ConcurrentDictionary<Tuple<ushort, Type>, TaskCompletionSource<MqttBasePacket>> _awaiters = new ConcurrentDictionary<Tuple<ushort, Type>, TaskCompletionSource<MqttBasePacket>>();
-        
-        public void Dispatch(Exception exception)
+
+        public async Task Dispatch(Exception exception)
         {
             foreach (var awaiter in _awaiters)
             {
-                Task.Run(() => awaiter.Value.TrySetException(exception)); // Task.Run fixes a dead lock. Without this the client only receives one message.
+                awaiter.Value.TrySetException(exception);
             }
 
-            _awaiters.Clear();
+            // await the tasks to complete
+            try
+            {
+                await Task.WhenAll(_awaiters.Values.Select(t => t.Task)).ConfigureAwait(false);
+            }
+            catch (Exception)
+            {
+                // swallow any exception that happens during exception handling to prevent
```

Comment:
why not just throw? At the minimum people should do proper exception handling but when they don't is should just crash and burn. If not, this will definitely result in super unwanted hidden errors.

Reply:
I am happy that you are raising concern about this line. Not having this in place may result however in collateral damage. As far as I am concerned this needs thorough insights by whoever is most familiar with this. @chkr1011, would you please have a look?

I do also see an issue about the previous `Task.Run` that was put in place to avoid deadlocks. A proper async/await application does not need `Task.Run` because it completes as soon as the task completes. On the contrary, when an async/await library like MQTTnet needs `Task.Run` to avoid deadlocks it is a strong indicator that there's something fishy going on somewhere. Something along the lines of:

* `Task t; t.GetAwaiter().GetResult()`
* `Task t; t.Wait()`

I propose to remove ANY `Task.Run` invocation and fix any emerging issues. This is roughly what I planned to summarize in a separate issue as already mentioned in #551 (more specifically [this comment](https://github.com/chkr1011/MQTTnet/pull/551#issuecomment-464443340)). Such a change probably requires a refactor of the alive/send/receive mechanisms to something like:

```csharp
private async Task RunAsync(CancellationToken token) {
    var awaitPacketsToSendCancellation = new CancellationTokenSource(); // wrap in using
    // add more cancellation token sources for each task
    Task awaitPacketsToSendTask = AwaitPacketsToSendAsync(awaitPacketsToSendCancellation.Token);
    Task receiveTask = ReceiveAsync(token);
    Task awaitCancellationTask = token.AwaitCancellationAsync(); // an extension method that creates a     TaskCompletionSource and sets the result when token.Register() is called
    Task awaitSendAliveCheckTask = TimeSpan.FromSeconds(5).AwaitSendAliveAsync(token); // an extension method that does roughly Task.Delay(timeSpan) and handles OperationCancelledException gracefully
    while (!token.IsCancellationRequested) {
    Task completedTask = await Task.WhenAny(sendTask, receiveTask, awaitCancellationTask, awaitSendAliveCheckTask);
    if (completedTask == awaitPacketsToSendTask) {
        var packetsToSend = await awaitPacketsToSendTask;
        await SendPacketsAsync(packetsToSend, token);
        awaitPacketsToSendTask = AwaitPacketsToSendAsync(token);
    }
    else if (completedTask == receiveTask) { /* process received packets and restart receiveTask */ }
    else if (completedTask == awaitSendAliveCheckTask) { /* check the elapsed time since the last successful send, eventually sending an alive and then start a new await send alive check task */ }
    else if (completedTask == awaitCancellationTask) { /* terminate gracefully if token.IsCancellationRequest == true by cancelling all child tokens and terminate when the connection etc was teared down */ }
    else { /* throw */ }
}
```

@ramonsmits What do you think?

Reply:
Task.Run must indeed not be used to resolve deadlock issues however there can be valid reasons for using it with CPU bound would still benefit from parallelism. However, this is applicable to the actual processing of the message not really the framework. 

To me is seems seems a bit strange to manage as individual tasks but I'm not familiar enough with this code base but the whole `else if` nesting doesn't feel really very clean.

Reply:
At the end of the day, if MQTTnet calls `Task.Run` the processing of messages ends happens in parallel across multiple threads even when the library has to maintain the absolute ordering of messages even across multiple topics.

Thanks for you valuable insights regarding the pseudo-code posted earlier. It may be possible to start several tasks and each task loops until it is cancelled. I have not implemented that pattern until now but only followed the above implementation pattern - mainly because it works. In any case the code should gracefully handle the following situation:

1. The broker sends a PUB
1. The PUB is received by the receive task
1. The receive task parses the message and calls the new api, allowing the application to processes the PUB; the code awaits until the async operation completes
1. the application code would then do several application logic stuff and eventually do several PUB by doing mqttClient.PublishAsync() and await that task to complete.
1. The await packets to send task would at this point complete and send the packets to be sent
1. Only when this completes, the processing of the PUB completes

An interesting question that emerges is: when should PUBACK for QoS 1 (or PUBREL for Qos 2) be sent? There are two candidates:

1. It be sent before the application code is allowed to handle the PUB
1. It is sent afterwards when the application code gracefully processed the PUB

The second solution makes the application re-entrant for the broker and that may not be desired. At the same time it would cause the broker to resend a PUB until the client gracefully handles the PUB.

Reply:
Lets not discuss multiple things here :-), I don't think the discussion on PUBACK belongs here. Shall we create a separate issue for that?